src/canister/user_index/src/api/canister_management/recycle_canisters.rs

```diff
@@ -0,0 +1,118 @@
+use std::time::SystemTime;
+
+use candid::Principal;
+use futures::StreamExt;
+use ic_cdk::{
+    api::management_canister::main::{self, CanisterIdRecord},
+    query, update,
+};
+use serde::Deserialize;
+use shared_utils::common::utils::permissions::is_reclaim_canister_id;
+use shared_utils::{
+    canister_specific::user_index::types::RecycleStatus,
+    common::{
+        types::wasm::{CanisterWasm, WasmType},
+        utils::{permissions::is_caller_controller, system_time::get_current_system_time},
+    },
+};
+
+use crate::{util::canister_management::reinstall_canister_wasm, CANISTER_DATA};
+
+#[query]
+pub fn get_recycle_status() -> RecycleStatus {
+    CANISTER_DATA.with_borrow(|canister_data| canister_data.recycle_status.clone())
+}
+
+#[update(guard = "is_reclaim_canister_id")]
+pub fn recycle_canisters(canister_ids: Vec<Principal>) {
+    // return if principal id is `rimrc-piaaa-aaaao-aaljq-cai`
+    // for a secondary measure to prevent accidental recycling of hotornot canisters
+    if ic_cdk::id() == Principal::from_text("rimrc-piaaa-aaaao-aaljq-cai").unwrap() {
+        return;
+    }
+
+    ic_cdk::spawn(async move {
+        let start = get_current_system_time();
+
+        let individual_user_template_canister_wasm = CANISTER_DATA.with_borrow(|canister_data| {
+            canister_data
+                .wasms
+                .get(&WasmType::IndividualUserWasm)
+                .unwrap()
+                .clone()
+        });
+
+        let futures = canister_ids.iter().map(|canister_id| {
+            recycle_canister(*canister_id, individual_user_template_canister_wasm.clone())
+        });
+
+        let stream = futures::stream::iter(futures).boxed().buffer_unordered(20);
+
+        let results = stream
+            .collect::<Vec<Result<Option<Principal>, (Principal, String)>>>()
+            .await;
+
+        // update recycle_status
+
+        let success_canisters = results
+            .iter()
+            .filter(|r| r.is_ok() && r.as_ref().unwrap().is_some())
+            .map(|r| r.as_ref().unwrap().unwrap())
+            .collect::<Vec<Principal>>();
+
+        let num_success = success_canisters.len();
```

Comment:
Instead of this, we can use `run_task_concurrently` method from utils.

Reply:
yeah saw that, but I want to return the result and do some actions. using result_callback seemed less intuitive for that